### DISCUSSION TITLE
vite_legacy_javascript_tag execution order inconsistent with vite_javascript_tag

### DISCUSSION THREAD

**Discussion by duprasa** [ElMassimo/vite_ruby#106] [Q&A] [answered]
### Description 📖

When using multiple endpoints on one page with `vite_legacy_javascript_tag` doesn't preserve execution order. I think this is because under the hood it just runs `System.import(document.getElementById('#{ id }')` behind the scenes which doesn't guarantee any execution order between `vite_legacy_javascript_tag` since they are handled async and will just run in completion order.

`vite_javascript_tag` does however behave as expected this is because `<script type="module">` will execute the endpoints in the order that they are defined in the html.

I know that ideally we would only have 1 endpoint per page, but thats not a viable option for us since we have a lot of legacy code :sweat_smile:  

### Reproduction 🐞

```
<head>
  <%= vite_javascript_tag "first_to_run" %>
  <%= vite_legacy_javascript_tag "first_to_run" %>
  <%= vite_javascript_tag "second_to_run" %>
  <%= vite_legacy_javascript_tag "second_to_run" %>
</head>
```

expected execution order: **first_to_run**, **second_to_run**
expected execution order for non-legacy browsers : **first_to_run**, **second_to_run**
expected execution order for legacy browsers : **first_to_run**, **second_to_run** or **second_to_run**, **first_to_run**

_Run `bin/rake vite:info` and provide the output:_

```
bin/vite present?: true
vite_ruby: 1.2.12
vite_rails_legacy: 2.0.12
rails: 4.2.11.1
node: v15.6.0
npm: 7.4.0
yarn: 1.22.4
bin/rake: No such file or directory - pnpm
pnpm: 
ruby: ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-linux]

installed packages:
rtings@ /home/sam/Workspace/rtings
├─┬ @vitejs/plugin-legacy@1.4.3
│ └── vite@2.4.2 deduped
├─┬ vite-plugin-ruby@2.0.4
│ └── vite@2.4.2 deduped
├─┬ vite-plugin-vue2@1.7.2
│ └── vite@2.4.2 deduped
└── vite@2.4.2

```
</details>

### Appreciation

Other than this issue, this gem has been awesome! It's really gonna help us out! Thanks :smile: :+1:

**Comment by ElMassimo** [comment]
Hi @duprasa!

As you mentioned, ideally you would only have one legacy entrypoint per page.

In order to control the order, you could change the [helper](https://github.com/ElMassimo/vite_ruby/blob/main/vite_plugin_legacy/lib/vite_plugin_legacy/tag_helpers.rb#L7-L14) contract from:

```ruby
def vite_legacy_javascript_tag(name, asset_type: :javascript)
```

to

```ruby
def vite_legacy_javascript_tag(*names, asset_type: :javascript)
```

and chain `System.import` promises:

```ruby
import_tag = content_tag(:script, nomodule: true) {
  <<-JS
    var srcs = #{vite_asset_paths.to_json.html_safe}
    var i = 0
    var promise = Promise.resolve()
    do {
      promise = promise.then(System.import(srcs[i]))
    } while (i++ < srcs.length)
  JS
}
```

This introduces additional complexity to the helper, so I'd like to hear from other users that share this requirement before updating the gem.

Please do try it out by overriding or creating a new helper based on this and let me know how it goes 😃

**Comment by ElMassimo** [comment, marked as answer]
It seems that when using the module polyfill, the order of `System.import` calls does not guarantee execution order.

A way to enforce execution order is to leverage the module system.

For example, in your case, create another entrypoint `combined_first_and_second.js`:

```js
import '~/entrypoints/first_to_run'
import '~/entrypoints/second_to_run'
```

and then reference this entrypoint in your pages:

```erb
<%= vite_javascript_tag "combined_first_and_second" %>
<%= vite_legacy_javascript_tag "combined_first_and_second" %>
```

<hr/>

If creating a shared entrypoint is not an option because the tags are being rendered in different views or partials, then there's really no guarantee to the execution order given the async nature of ES modules.

**Comment by duprasa** [comment]
Thanks for all the help, I'll be looking at this problem again in a couple days, you gave me a few interesting ideas. I'll let you know if I come up with anything interesting. It does seem my options are quite limited. Might be able to convince the boss to drop IE instead 🤞 😄
